### ISSUE TITLE
segfault in GroupedAggregateHashTable::Resize 

### ISSUE THREAD

**Issue by Diskein** [duckdb/duckdb#8031]
### What happens?

I've recently faced with segfault on my dataset, which is pretty big, it's around 14 billions of records with two integer's columns. I believe my setup isn't common because it's a server with 800+ GB RAM, and Intel Xeon CPU @ 2.60GHz.

Query:
```SQL
CREATE TABLE edges AS
SELECT DISTINCT "From", "To"
FROM
(
    SELECT "From", "To"
    FROM edges1
    UNION
    SELECT "To" AS "From", "From" AS "To"
    FROM edges1
)
```

It happens at `duckdb::GroupedAggregateHashTable::Resize<duckdb::aggr_ht_entry_64>` at `/opt/duckdb/src/execution/aggregate_hashtable.cpp:210` (v0.7.1, tried v0.8.1 - the same).
<details><summary>Full stack trace</summary>
<p>

```
received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fe9428ff6c0 (LWP 173)]
0x00007fe9a838e2ba in ?? () from /usr/lib/libc.so.6
(gdb) bt 100
#0  0x00007fe9a838e2ba in ?? () from /usr/lib/libc.so.6
#1  0x00007fe9a9862cf4 in duckdb::GroupedAggregateHashTable::Resize<duckdb::aggr_ht_entry_64> (this=this@entry=0x7f8a4d5df410, size=536870912) at /opt/duckdb/src/execution/aggregate_hashtable.cpp:210
#2  0x00007fe9a986715d in duckdb::GroupedAggregateHashTable::FindOrCreateGroupsInternal<duckdb::aggr_ht_entry_64> (this=0x7f8a4d5df410, groups=..., group_hashes=..., addresses=..., new_groups_out=...) at /opt/duckdb/src/execution/aggregate_hashtable.cpp:338
#3  0x00007fe9a986b75d in duckdb::GroupedAggregateHashTable::FlushMove (this=this@entry=0x7f8a4d5df410, state=..., source_addresses=..., source_hashes=..., count=2048) at /opt/duckdb/src/execution/aggregate_hashtable.cpp:519
#4  0x00007fe9a986be8f in operator() (page_nr=<optimized out>, page_offset=<optimized out>,
    ptr=0x7fd008d96c30 "\227K\200\305\004\241b\202\a\260 =^G\352Ņ \235!G\a\330\t\027K\200\305\004@2\b\b\\\344q\344#*\324]\025\206\025\v\n\270 \aK\200\305\004\212g\253\t-\366y\260\004j\217\a\v\027\001\317\027|\031\347K\200\305\004}\3008\006\027\245Gf\212\n`\310\003q\t\265\004\251\"\367K\200\305\004\350'\205\b\205\374_1\bJ\3122\004\221\006\017\r\t'/\203\303|\bk\203*\002f:\002\217A\252\277\027 G&\340\n\243\0057\203\303|\bt\f\005\n\\\272B$\363*\222^", __closure=<synthetic pointer>) at /opt/duckdb/src/execution/aggregate_hashtable.cpp:552
#5  duckdb::GroupedAggregateHashTable::PayloadApply<duckdb::GroupedAggregateHashTable::Combine(duckdb::GroupedAggregateHashTable&)::<lambda(duckdb::idx_t, duckdb::idx_t, duckdb::data_ptr_t)> > (fun=..., this=0x7fe92fcbc400) at /opt/duckdb/src/execution/aggregate_hashtable.cpp:96
#6  duckdb::GroupedAggregateHashTable::Combine (this=0x7f8a4d5df410, other=...) at /opt/duckdb/src/execution/aggregate_hashtable.cpp:545
#7  0x00007fe9a986e513 in duckdb::RadixAggregateFinalizeTask::FinalizeHT (radix=10, gstate=...) at /opt/duckdb/src/execution/radix_partitioned_hashtable.cpp:278
#8  duckdb::RadixAggregateFinalizeTask::ExecuteTask (this=0x7fb49a6aef70, mode=<optimized out>) at /opt/duckdb/src/execution/radix_partitioned_hashtable.cpp:286
#9  0x00007fe9a991fc7f in duckdb::ExecutorTask::Execute (this=0x7fb49a6aef70, mode=<optimized out>) at /opt/duckdb/src/parallel/executor_task.cpp:17
#10 0x00007fe9a991cb27 in duckdb::TaskScheduler::ExecuteForever (this=0x7fe9a7c293f0, marker=0x7fe9a7c341f8) at /opt/duckdb/src/parallel/task_scheduler.cpp:135
#11 0x00007fe9a8517943 in std::execute_native_thread_routine (__p=0x7fe9a6b177a0) at /usr/src/debug/gcc/gcc/libstdc++-v3/src/c++11/thread.cc:104
#12 0x00007fe9a82ae44b in ?? () from /usr/lib/libc.so.6
#13 0x00007fe9a8331d64 in clone () from /usr/lib/libc.so.6

```
</p>
</details> 

It's a memset call on buffer's pointer. It seems that something wrong with allocation inside the buffer_manager and we get spoiled pointer (not null). The fail happens when the hashmap is trying to grow up to 4GiB

```shell
(gdb) info locals
byte_size = 4294967296
hashes_arr = 0x0
(gdb) info args
this = 0x7f8a4d5df410
size = 536870912
(gdb) p hashes_hdl.IsValid()
$26 = true
```

ASAN didn't help. 

```shell
Address 0x7fa2d5b28000 is a wild pointer inside of access range of size 0x000100000000.
SUMMARY: AddressSanitizer: unknown-crash /usr/src/debug/gcc/gcc/libsanitizer/sanitizer_common/sanitizer_common_interceptors.inc:871 in __interceptor_memset
Shadow bytes around the buggy address:
  0x7fa2d5b27d80: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x7fa2d5b27e00: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x7fa2d5b27e80: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x7fa2d5b27f00: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x7fa2d5b27f80: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x7fa2d5b28000:[fe]fe fe fe fe fe fe fe fe fe fe fe fe fe fe fe
  0x7fa2d5b28080: fe fe fe fe fe fe fe fe fe fe fe fe fe fe fe fe
  0x7fa2d5b28100: fe fe fe fe fe fe fe fe fe fe fe fe fe fe fe fe
  0x7fa2d5b28180: fe fe fe fe fe fe fe fe fe fe fe fe fe fe fe fe
  0x7fa2d5b28200: fe fe fe fe fe fe fe fe fe fe fe fe fe fe fe fe
  0x7fa2d5b28280: fe fe fe fe fe fe fe fe fe fe fe fe fe fe fe fe
```

### To Reproduce

CREATE TABLE edges1("From" INTEGER, "To" INTEGER);
-- generate dataset with 7kkk records, from query
loop 7000000000
INSERT INTO edges1 SELECT (random() * 300000000)::INTEGER as "From", (random()*300000000)::INTEGER as "To";
CREATE TABLE edges AS
SELECT DISTINCT "From", "To"
FROM
(
    SELECT "From", "To"
    FROM edges1
    UNION
    SELECT "To" AS "From", "From" AS "To"
    FROM edges1
);

### OS:

Arch Linux, 20230618.0.159075, x86

### DuckDB Version:

0.7.1

### DuckDB Client:

C++

### Full Name:

Denis Kalantaevskii

### Affiliation:

None

### Have you tried this on the latest `master` branch?

- [X] I agree

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] I agree

**Comment by lnkuiper** [comment]
Hi @Diskein, thanks for the bug report. I cannot reproduce this as I don't have access to a machine with that much memory.

However, we have just merged #8475, which reworks aggregations. I think this should fix your problem, as the way we resize the HT has changed.

**Comment by Diskein** [comment]
Hey @lnkuiper, thank you, I will try to check it.
I'll keep you posted.
